plugins/inputs/logfile/logfile.go

```diff
@@ -205,6 +171,13 @@ func (t *LogFile) FindLogSrc() []logs.LogSrc {
 				continue
 			}
 
+			if t.numUsedFds.GetLimit() != 0 {
+				if ok := t.numUsedFds.Acquire(defaultTimeoutToAcquire); !ok {
+					t.Log.Debugf("Cannot increase counter of used file descriptors")
+				}
+			}
```

Comment:
The semaphore code was really well done!

I don't think it needs to be used though.
It would be simpler and use less memory to have a single counter (integer instead of channel) of the open files (aka files being tailed). Then just log the counter's current value periodically.

It is undesirable to change behavior here and block for up to 1 second.

Reply:
For the information: there were others use semaphore as a counter (XSync, Marusama). However, after comparing the performance between them, I would say this version (or similar to DropBox), this would be the best option(for now) when decide the file descriptors as a upper bound.
```
BenchmarkSemaphore_AcquireReleaseUnderLimitSimple-12          19336089                62.26 ns/op            0 B/op          0 allocs/op
BenchmarkSemaphore_AcquireReleaseUnderLimit-12                    1652306               726.5 ns/op             0 B/op          0 allocs/op
BenchmarkSemaphore_AcquireReleaseOverLimit-12                     32737             37202 ns/op               0 B/op          0 allocs/op
BenchmarkMarusamaSemaphore_AcquireReleaseUnderLimitSimple-12                         6038138               171.7 ns/op            96 B/op          1 allocs/op
BenchmarkMarusamaSemaphore_AcquireReleaseUnderLimit-12                                 479114              2268 ns/op             960 B/op         10 allocs/op
BenchmarkMarusamaSemaphore_AcquireReleaseOverLimit-12                                   50464             24728 ns/op            9600 B/op        100 allocs/op
BenchmarkAbiosoftSemaphore_AcquireReleaseUnderLimitSimple-12                 8135151               145.0 ns/op             0 B/op          0 allocs/op
BenchmarkAbiosoftSemaphore_AcquireReleaseUnderLimit-12                          432472              2731 ns/op               0 B/op          0 allocs/op
BenchmarkAbiosoftSemaphore_AcquireReleaseOverLimit-12                           51364             27126 ns/op               0 B/op          0 allocs/op
BenchmarkPivotalGolangSemaphore_AcquireReleaseUnderLimitSimple-12            3358042               355.3 ns/op           128 B/op          2 allocs/op
BenchmarkPivotalGolangSemaphore_AcquireReleaseUnderLimit-12                      149810              7881 ns/op            1200 B/op         20 allocs/op
BenchmarkPivotalGolangSemaphore_AcquireReleaseOverLimit-12                      14558             83574 ns/op           12000 B/op        200 allocs/op
BenchmarkXSyncSemaphore_AcquireReleaseUnderLimitSimple-12                   40687284                27.22 ns/op            0 B/op          0 allocs/op
BenchmarkXSyncSemaphore_AcquireReleaseUnderLimit-12                             1000000              1060 ns/op               0 B/op          0 allocs/op
BenchmarkXSyncSemaphore_AcquireReleaseOverLimit-12                              17275             69725 ns/op           15994 B/op        299 allocs/op
```
However, the ` 1 second` is only an arbitrary value and it could be more or less. But as long as it within acceptable range (e.g do customer cares about the show number of monitored files being late in `1 second` or not? Maybe its exchange between source of truth and the fastness a little bit.

Reply:
Maybe I'm wrong but according to the `syns/atomic`
```
Package atomic provides low-level atomic memory primitives useful for implementing synchronization algorithms.
These functions require great care to be used correctly. Except for special, low-level applications, synchronization is better done with channels or the facilities of the sync package. 
```
So I'm still considering between the two. However, will compare performance before giving my final decision.

Reply:
Compare a little bit when doing `atomic counter with mutex` 
```
BenchmarkSemaphore_Acquire_Release_under_limit-12           9673            124611 ns/op               0 B/op          0 allocs/op
BenchmarkAtomicCounterWithOutLock_Under-12                            19537             53280 ns/op               0 B/op          0 allocs/op
BenchmarkAtomicCounterWithLock_Under-12                              7071            247445 ns/op              18 B/op          0 allocs/op
```
So maybe I'm still standing by with using semapahore instead

Reply:
Sorry to derail, I wasn't as concerned about the implementation of counter vs semaphore.
Agreed that an atomic counter (no lock needed) is simpler and would suffice here.

More concerned that you are introducing a blocking function call on `Acquire()`.
I initially thought this PR would just add a variable to track the count of open files AND then periodically log it.
Not to check the counter value and change behavior based on the counter value.

Reply:
Yeah, I know it does not block indefinitely, but I consider it "blocking" even if it is blocks for 100ms.

*** I don't think it should block at all. I propose setting `defaultTimeoutToAcquire` to 0, or just removing the timeout parameter from `Acquire()`.***

Reply:
Agreed. The timeout stops it from blocking indefinitely but is still potentially paused for up to 100ms. That's not great